File: 24-swap-nodes-in-pairs/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* swapPairs(ListNode* head) {
    if (head == NULL || head->next == NULL) {
      return head;
    }

    ListNode* d = new ListNode(0, head);
    ListNode* a = d;
    ListNode* b = d->next;

    while (b != NULL && b->next != NULL) {
      a->next = b->next;
      b->next = a->next->next;
      a->next->next = b;
      b = b->next;
      a = a->next->next;
    }

    return d->next;
  }
};

File: 24-swap-nodes-in-pairs/go/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	d := &ListNode{Val: 0, Next: head}
	a := d
	b := d.Next

	for b != nil && b.Next != nil {
		a.Next = b.Next
		b.Next = a.Next.Next
		a.Next.Next = b
		a = a.Next.Next
		b = b.Next
	}

	return d.Next
}
